File: erlang/tokencontrol.erl
-module(tokencontrol).
-compile(export_all).
%-export([setU/0, tickTime/0, recvT/1, getT/, mustProc/0, makeToken/2, getListaAltas/2, getListaBajas/2]).

% Este es el modulo que encapsula el manejo del token
% Funciones
% tickTime : () -> Int (devuelve el tiempo que tarda el token en ser procesado)
% recvT : Token -> ok (notifica al modulo que se recibio el Token y lo almacena)
% getT : () -> Token (devuelve el Token almacenado)
% mustProc() : () -> true | false (indica si se debe procesar o no el Token)
% makeToken : [ {NameFile, WorkerId} ],  [ {NameFile,WorkerId} ] -> Token
% getListaAltas : Token -> [ {NameFile,WorkerId} ]
% getListaBajas : Token -> [ {NameFile,WorkerId} ]

loop( T, F ) ->

%    io:format("tokencontrol: ~p ~p~n",[T,F]), %DEBUG
    receive
        { P, getT }  ->  P ! {tokencontrolserverResponse, T },
                           loop( nothing , false ) ;
        {_, recvT, TT}   ->  loop( TT, true ) ;
        {P, mustProc}   -> P ! {tokencontrolserverResponse, F},
                                loop(T,F) ;
        _ -> error("esto no deberia suceder 666")
    end.
setUp() -> register( tokencontrolserver, spawn(?MODULE,loop,[nothing,false]) ).
tickTime() -> 10.
recvT_aux(TT) -> 
            receive after tickTime() -> ok end,
            tokencontrolserver ! {self(), recvT, TT }, 
            ok.
recvT(TT) -> spawn(?MODULE,recvT_aux,[TT]), ok.
getT() -> tokencontrolserver ! {self(), getT }, 
                       receive {tokencontrolserverResponse, X} -> X end.
mustProc()-> tokencontrolserver ! {self(), mustProc }, 
                       receive {tokencontrolserverResponse, X} -> X end.

makeToken(LA,LB) -> {LA,LB}.
getListaAltas({LA,_}) -> LA.
getListaBajas({_,LB}) -> LB.

File: cpp/tokenControl.cpp
#include "tokenControl.h"

vector< pair<string,WorkerId> > token::getListaAltas() {
	vector< pair<string,WorkerId> > res = lc;
	lc.clear();
	return res;
}

vector< pair<string,WorkerId> > token::getListaBajas() {
	vector< pair<string,WorkerId> > res = ld;
	ld.clear();
	return res;
}

token::token() {}
token::token(const vector< pair<string,WorkerId> >& lcp, vector< pair<string,WorkerId> >& ldp) {
	lc = lcp;
	ld = ldp;
}



int tokenControl::tickTime() { return TOKEN_MICROSECONDS_TIME; }

void tokenControl::recvT(const token& nt) {
	assert(!valid);
	valid=true;
	trec=clock();
	t=nt;
}
token tokenControl::getT() {
	assert(valid);
	valid=false;
	return t;
}
bool tokenControl::mustProc() {
	if (!valid) return false;
	double seconds = double(clock()-trec) / CLOCKS_PER_SEC;
	return seconds > double(TOKEN_MICROSECONDS_TIME) / 1000000. ;
}

tokenControl::tokenControl() { valid=false; }


